File: com.trollworks.gcs/src/com/trollworks/gcs/modifier/EquipmentModifierWeightType.java
/*
 * Copyright ©1998-2020 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.modifier;

import com.trollworks.gcs.utility.I18n;
import com.trollworks.gcs.utility.units.WeightUnits;

/** Describes how a {@link EquipmentModifier}'s weight is applied. */
public enum EquipmentModifierWeightType {
    /**
     * Modifies the original value stored in the equipment. Can be a ±value or a ±% value. Examples:
     * '+5 lb', '-5 lb', '+10%', '-10%'
     */
    TO_ORIGINAL_WEIGHT(ModifierWeightValueType.ADDITION, ModifierWeightValueType.PERCENTAGE_ADDER) {
        @Override
        public String toShortString() {
            return I18n.Text("to original weight");
        }

        @Override
        public String toString() {
            return toShortString() + I18n.Text(" (e.g. '+5 lb', '-5 lb', '+10%', '-10%')");
        }
    },
    /**
     * Modifies the base cost. Can be a ±value, a ±% value, or a multiplier. Examples: '+5 lb',
     * '-5 lb', 'x10%', 'x3', 'x2/3'
     */
    TO_BASE_WEIGHT(ModifierWeightValueType.ADDITION, ModifierWeightValueType.PERCENTAGE_MULTIPLIER, ModifierWeightValueType.MULTIPLIER) {
        @Override
        public String toShortString() {
            return I18n.Text("to base weight");
        }

        @Override
        public String toString() {
            return toShortString() + I18n.Text(" (e.g. '+5 lb', '-5 lb', 'x10%', 'x3', 'x2/3')");
        }
    },
    /**
     * Modifies the final base cost. Can be a ±value, a ±% value, or a multiplier. Examples:
     * '+5 lb', '-5 lb', 'x10%', 'x3', 'x2/3'
     */
    TO_FINAL_BASE_WEIGHT(ModifierWeightValueType.ADDITION, ModifierWeightValueType.PERCENTAGE_MULTIPLIER, ModifierWeightValueType.MULTIPLIER) {
        @Override
        public String toShortString() {
            return I18n.Text("to final base weight");
        }

        @Override
        public String toString() {
            return toShortString() + I18n.Text(" (e.g. '+5 lb', '-5 lb', 'x10%', 'x3', 'x2/3')");
        }
    },
    /**
     * Modifies the final cost. Can be a ±value, a ±% value, or a multiplier. Examples: '+5 lb',
     * '-5 lb', 'x10%', 'x3', 'x2/3'
     */
    TO_FINAL_WEIGHT(ModifierWeightValueType.ADDITION, ModifierWeightValueType.PERCENTAGE_MULTIPLIER, ModifierWeightValueType.MULTIPLIER) {
        @Override
        public String toShortString() {
            return I18n.Text("to final weight");
        }

        @Override
        public String toString() {
            return toShortString() + I18n.Text(" (e.g. '+5 lb', '-5 lb', 'x10%', 'x3', 'x2/3')");
        }
    };

    private final ModifierWeightValueType[] mPermittedTypes;

    EquipmentModifierWeightType(ModifierWeightValueType... permittedTypes) {
        mPermittedTypes = permittedTypes;
    }

    public abstract String toShortString();

    public ModifierWeightValueType determineType(String text) {
        ModifierWeightValueType mvt = ModifierWeightValueType.determineType(text);
        return allowed(mvt) ? mvt : mPermittedTypes[0];
    }

    public Fraction extractValue(String text, boolean localized) {
        return determineType(text).extractFraction(text, localized);
    }

    public String format(String text, WeightUnits defUnits, boolean localized) {
        ModifierWeightValueType mvt = determineType(text);
        String                  str = mvt.format(mvt.extractFraction(text, localized), localized);
        if (mvt == ModifierWeightValueType.ADDITION) {
            str += " " + ModifierWeightValueType.extractUnits(text, defUnits).getAbbreviation();
        }
        return str;
    }

    private boolean allowed(ModifierWeightValueType mvt) {
        for (ModifierWeightValueType allowed : mPermittedTypes) {
            if (allowed == mvt) {
                return true;
            }
        }
        return false;
    }
}

File: model/gurps/equipment_modifier_weight_type.go
/*
 * Copyright ©1998-2022 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package gurps

import (
	"fmt"

	"github.com/richardwilkes/gcs/v5/model/fxp"
	"github.com/richardwilkes/gcs/v5/model/gurps/measure"
)

// StringWithExample returns an example along with the normal String() content.
func (enum EquipmentModifierWeightType) StringWithExample() string {
	return fmt.Sprintf("%s (e.g. %s)", enum.String(), enum.AltString())
}

// Permitted returns the permitted EquipmentModifierCostValueType values.
func (enum EquipmentModifierWeightType) Permitted() []EquipmentModifierWeightValueType {
	if enum.EnsureValid() == OriginalEquipmentModifierWeightType {
		return []EquipmentModifierWeightValueType{AdditionEquipmentModifierWeightValueType, PercentageAdderEquipmentModifierWeightValueType}
	}
	return []EquipmentModifierWeightValueType{AdditionEquipmentModifierWeightValueType, PercentageMultiplierEquipmentModifierWeightValueType, MultiplierEquipmentModifierWeightValueType}
}

// DetermineModifierWeightValueTypeFromString examines a string to determine what type it is, but restricts the result to
// those allowed for this EquipmentModifierWeightType.
func (enum EquipmentModifierWeightType) DetermineModifierWeightValueTypeFromString(s string) EquipmentModifierWeightValueType {
	mvt := LastEquipmentModifierWeightValueType.FromString(s)
	permitted := enum.Permitted()
	for _, one := range permitted {
		if one == mvt {
			return mvt
		}
	}
	return permitted[0]
}

// ExtractFraction from the string.
func (enum EquipmentModifierWeightType) ExtractFraction(s string) fxp.Fraction {
	return enum.DetermineModifierWeightValueTypeFromString(s).ExtractFraction(s)
}

// Format returns a formatted version of the value.
func (enum EquipmentModifierWeightType) Format(s string, defUnits measure.WeightUnits) string {
	t := enum.DetermineModifierWeightValueTypeFromString(s)
	result := t.Format(t.ExtractFraction(s))
	if t == AdditionEquipmentModifierWeightValueType {
		result += " " + measure.TrailingWeightUnitsFromString(s, defUnits).String()
	}
	return result
}
